### ISSUE TITLE
Resolvers: Create tests for removeOrganization.js

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#554]
The Talawa-API code base needs to be 100% reliable. This means we need to have 100% test code coverage.

Tests need to be written for file `lib/resolvers/organization_mutations/removeOrganization.js`

- We will need the API to be refactored for all methods, classes and/or functions found in this file for testing to be correctly executed.
- When complete, all all methods, classes and/or functions in the refactored file will need to be tested. These tests must be placed in a
single file with the name `tests/resolvers/organization_mutations/removeOrganization.spec.js`. You may need to create the appropriate directory structure to do this.

### IMPORTANT:
Please refer to the parent issue on how to implement these tests correctly:

- https://github.com/PalisadoesFoundation/talawa-api/issues/490

### PR Acceptance Criteria

- When complete this file must show **100%** coverage when merged into the code base. This will be clearly visible when you submit your PR.
- [The current code coverage for the file can be found here](https://codecov.io/gh/PalisadoesFoundation/talawa-api/tree/develop/lib/resolvers/organization_mutations/). If the file isn't found in this directory, or there is a 404 error, then tests have not been created.
- The PR will show a report for the code coverage for the file you have added. You can use that as a guide.

**Comment by ebrahim354** [comment]
I want to work on this please

**Comment by palisadoes** [comment]
I'm going to have to revert the code again. It's causing other tests to fail.

- https://github.com/PalisadoesFoundation/talawa-api/runs/5759635905?check_suite_focus=true

Someone else is going to have to work on this.

**Comment by palisadoes** [comment]
Here is the error we are getting:

![image](https://user-images.githubusercontent.com/16875803/160909538-10a96568-4271-41bd-924b-d5545b30cdef.png)

**Comment by ebrahim354** [comment]
The tests are failing because they relay on the files: organizationExists, adminCheck that are changed on this PR https://github.com/PalisadoesFoundation/talawa-api/pull/653 the pull requester should fix any affected files after the refactor of the code, but I can easily fix it if you want, also I still want to work on it please.

**Comment by ebrahim354** [comment]
He changed the format of errors thrown without fixing the rest of the tests that relay on it.

**Comment by palisadoes** [comment]
This is getting confusing. I've reassigned the issue to you. Please resubmit the PR.

**Comment by xoldyckk** [comment]
@palisadoes why are we using two approaches to test the resolvers? This is creating problems. Please decide on one approach and let us know. Either we run tests on the running server or we run the tests on just the resolver function. For first approach we'll have to run tests against a running server with a test-database. For second approach we just need to run a test-database. @ebrahim354 is using the first approach and I'm using the second approach. I'm only using the second approach because I saw many tests were already written that way inside `tests/resolvers`. Personally I too like the first approach because with that we won't have to refactor the resolvers to throw different errors in testing and production environment because of `i18n initialization error`.

**Comment by palisadoes** [comment]
I'd prefer to have tests that mimic a production system as much as possible, using mocks and if absolutely necessary a live server. @DMills27  and @JamaicanFriedChicken are working on an approach. Let's see what they have to say.

**Comment by ebrahim354** [comment]
> @palisadoes why are we using two approaches to test the resolvers? This is creating problems. Please decide on one approach and let us know. Either we run tests on the running server or we run the tests on just the resolver function. For first approach we'll have to run tests against a running server with a test-database. For second approach we just need to run a test-database. @ebrahim354 is using the first approach and I'm using the second approach. I'm only using the second approach because I saw many tests were already written that way inside `tests/resolvers`. Personally I too like the first approach because with that we won't have to refactor the resolvers to throw different errors in testing and production environment because of `i18n initialization error`.

I've finished fixing my block and unblock to test the resolver directly that should make both of us pass the test cases then the main developers should decide on a standard approach to be used from now on  so there is no conflicts between PRs.

**Comment by palisadoes** [comment]
> > @palisadoes why are we using two approaches to test the resolvers? This is creating problems. Please decide on one approach and let us know. Either we run tests on the running server or we run the tests on just the resolver function. For first approach we'll have to run tests against a running server with a test-database. For second approach we just need to run a test-database. @ebrahim354 is using the first approach and I'm using the second approach. I'm only using the second approach because I saw many tests were already written that way inside `tests/resolvers`. Personally I too like the first approach because with that we won't have to refactor the resolvers to throw different errors in testing and production environment because of `i18n initialization error`.
> 
> I've finished fixing my block and unblock to test the resolver directly that should make both of us pass the test cases then the main developers should decide on a standard approach to be used from now on so there is no conflicts between PRs.

This may not be the right approach long term. Let's see what the other mentors have to say. They are researching options. I've contacted them off this thread. They should respond shortly.

**Comment by ebrahim354** [comment]
> I'd prefer to have tests that mimic a production system as much as possible, using mocks and if absolutely necessary a live server. @DMills27 and @JamaicanFriedChicken are working on an approach. Let's see what they have to say.

The first approach is better for mocking a server and also there is less code changes as  @xoldyckk said but the test coverage is not counted properly and the second approach is less code and easier to write so you guys should decide.

**Comment by DMills27** [comment]
The first approach is standardised for both the test and production server so we should use that one. Can you explain why there would not be sufficient test coverage in the first approach? Having a test database, as in the second approach, doesn't necessarily suggest that it is the more comprehensive option.

**Comment by ebrahim354** [comment]
I previously submitted a PR: #646 with tests for the block and unblock user resolvers testing all the possibilities but the coverage was 0 also when I rewritten the tests using the resolver directly I found a bug that I didn't find with the first trial. the coverage problem I explained it  on https://github.com/PalisadoesFoundation/talawa-api/issues/578 but I'm not sure if that is the reason though.

**Comment by palisadoes** [comment]
@ebrahim354 

- What was the nature of the test you found with the resolver directly versus via the API? 
- Could it have been discovered with more thorough testing via the API? 
- If not, is it practical to use both approaches, or is it one or the other?
- The preferred approach is to have a standardized approach. Do you know of any way to make codecov work with APIs? Do you know of any other repos that have overcome this issue effectively

@xoldyckk It would be good to hear your thoughts on this. 

This will impact the future development of the repo and we need to make the right decision now.

**Comment by ebrahim354** [comment]
- The resolver was saving a whole organization object inside of a user record instead of just saving the org._id and use populate on it later which is problematic and that object(a mongoose object) was so big it didn't even fit in a console.log.
- It couldn't have been discovered with the API approach because graphql only sends specific fields after getting records from the database.
- It's possible to use both: for example testing via the api for all possible cases and just do on unit test only on a correct case and assert on the expected output but the coverage problem is still present.
- The problems is: jest collects coverage only on it's own process, what that means is if we run the test-server on a process (one terminal for simplicity) and then run jest on the other one jest doesn't see the other process, the only solution is to call the api from with in jest and that would be using a beforeAll() that starts the server tests it and then an afterAll() that stops the server and so on before every single test file,
and that would cause the issue of reserved ports because jest uses multible threads to run many tests at the same time but this issue can be solved easily using port 0 for example.

**Comment by palisadoes** [comment]
I spoke with Dominic. We will need to go with the live API approach. We'll need to figure out a way to evaluate the code coverage using this method, possibly using another tool

**Comment by ebrahim354** [comment]
So my reverted PRs should be ok the way they are or should I add some changes?.
Also this page might be helpful : 
https://www.apollographql.com/docs/apollo-server/testing/testing

**Comment by palisadoes** [comment]
Yes they should be resubmitted. 

- How do we prevent future PRs from breaking? Do we add the testing regimen to the CONTRIBUTING.md file? 
- Will past tests break with your PRs addition?

We'll need to covert existing tests to this approach too.

**Comment by ebrahim354** [comment]
usually my process is after writing code and commiting it locally I run all the tests locally and if they pass I submit if they don't I don't submit so the PR breaking is the responsibility of the requester because he is the one who changed the passing tests.
this particular PR https://github.com/PalisadoesFoundation/talawa-api/pull/653 didn't break the system he just changed the format that the new tests expect to be there without fixing the tests and making them pass.
about converting all the tests in my opinion it's better to choose a standard way of testing for future tests only then after the system is fully tested think about changing the old ones, again this is just an opinion the main developers should know what is best.

**Comment by xoldyckk** [comment]
> @ebrahim354
> 
> * What was the nature of the test you found with the resolver directly versus via the API?
> * Could it have been discovered with more thorough testing via the API?
> * If not, is it practical to use both approaches, or is it one or the other?
> * The preferred approach is to have a standardized approach. Do you know of any way to make codecov work with APIs? Do you know of any other repos that have overcome this issue effectively
> 
> @xoldyckk It would be good to hear your thoughts on this.
> 
> This will impact the future development of the repo and we need to make the right decision now.

Since we are not mocking the database I think integration test corresponds the best with nature of resolver tests currently implemented. Though this is expected because the project is not following the best practices to easily implement unit or integration tests. For example this is very easy to do in a Nest.js project because everything is injected into the child components through dependency injection and every individual component can easily be unit/integration tested at any level. Since we're using apollo server [link](https://www.apollographql.com/docs/apollo-server/testing/testing/#testing-using-executeoperationl), they've made it easy to do integration tests without invoking the whole http server. Though `i18n not initialized` error will kick in again if we use this approach.

The nature of currently implemented API tests corresponds more to end to end testing.

It is very practical to use both kinds of approaches. For unit tests we use mock all internal/external services and test components in isolation. For integration tests we use mocks wherever possible and use real services if absolutely necessary. For end to end tests we simulate a real production-like environment inside our tests. Though it also depends on the size of our project, team, deadlines etc. It could be overkill to include every kind of test depending on those factors. I've been reading about testing in nodejs(server) ecosystem. And I'm still confused what they mean by integration tests and e2e tests. Usually there's not a clear distinction. It makes more sense in the frontend world where there are specialized tools like `cypress` specifically for e2e tests. 

The code coverage issue can easily be solved if the server we're testing against is invoked within the same process the test is running. This will require separating the server initialization and server definition logic into separate files.This way we can just import the httpServer and invoke it inside our tests and do other things like starting/stopping server instance within the tests/ randomizing the ports to avoid collision between tests etc.

**Comment by palisadoes** [comment]
@DMills27 @JamaicanFriedChicken Please respond. I'd like to get a definitive strategy so we can apply it to the PRs in the queue. The proposal to two both is interesting, especially if we provide guidance in the test issues. 

@ebrahim354 If we use mocks, does this mean that we won't need to check the ENV variable to simplify the code?

**Comment by ebrahim354** [comment]
If mocks means using Apollo's [executeOperation](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#executeoperation) then no need to check the ENV variable and the `i18n not initialized` will not occur since we are testing Apollo's interface itself with graphql queries.

**Comment by palisadoes** [comment]
This looks like the approach we should use. @DMills27 @JamaicanFriedChicken Please comment. 

@ebrahim354 Please prepare an issue to update CONTRIBUTING.md to describe the process you are proposing. We need to have others follow this methodology and direct them to the updated markdown when their tests fail. I'm assuming the issue will also need to include a method to create a blank DB etc. to make this all work.

**Comment by DMills27** [comment]
I agree that this approach suggested by @xoldyckk looks the most promising. I going to create a separate issue with the details you've proposed and I suggest you create a toy example for the purpose of verifying the proposed method and examining any hiccups.

**Comment by DMills27** [comment]
The new issue can be found here: https://github.com/PalisadoesFoundation/talawa-api/issues/660

**Comment by palisadoes** [comment]
I've been thinking about our resolver testing issues for some time. We have dozens of resolvers that require testing and it's just not scalable.

- Could we consider converting our resolver architecture to use a standardized filtering methodology like dGraph? 
- How easy would that be for us to consider? 
- Could it be done over the summer?

**Comment by palisadoes** [comment]
@sandeep-v1404, @sumitra19jha, @JamaicanFriedChicken  What do you think?

**Comment by arnab15** [comment]
> > @ebrahim354
> > 
> > * What was the nature of the test you found with the resolver directly versus via the API?
> > * Could it have been discovered with more thorough testing via the API?
> > * If not, is it practical to use both approaches, or is it one or the other?
> > * The preferred approach is to have a standardized approach. Do you know of any way to make codecov work with APIs? Do you know of any other repos that have overcome this issue effectively
> > 
> > @xoldyckk It would be good to hear your thoughts on this.
> > This will impact the future development of the repo and we need to make the right decision now.
> 
> Since we are not mocking the database I think integration test corresponds the best with nature of resolver tests currently implemented. Though this is expected because the project is not following the best practices to easily implement unit tests. Business layer, service layer, database layer are all implented in the same scope. For example this is very easy to do in a Nest.js project because everything is injected into the child components through dependency injection and every individual component can easily be unit/integration tested at any level because those dependencies can be mocked or instantiated according to needs in testing environement. Since we're using apollo server, they've made it easy to do integration tests without invoking the whole http server through this approach :- [link](https://www.apollographql.com/docs/apollo-server/testing/testing/#testing-using-executeoperation). Though `i18n not initialized` error will kick in if we use this approach.
> 
> The nature of currently implemented API tests corresponds more to end to end testing.
> 
> It is very practical to use both kinds of approaches. For unit tests we use mock all internal/external services and test components in isolation. The integration test is done on a module(collection of files) which reflects a self-contained unit of our project. Something that doesn't depend on other modules and has every functionality it needs baked into it. For this we use mocks wherever possible and use real services if absolutely necessary. For end to end tests we simulate a real production-like environment inside our tests. Though it also depends on the size of our project, team, deadlines etc. It could be overkill to include every kind of test depending on those factors. I've been reading about testing in nodejs(server) ecosystem and I'm still confused what they mean by integration tests and e2e tests. Usually there's not a clear distinction. Some write tests making request to the running http server and call it integration test. Some call it e2e test. Honestly it makes more sense in the frontend world where there are specialized tools like `cypress` specifically for e2e tests.
> 
> The code coverage issue for API tests can be solved if the server we're testing against is invoked within the same process the test is running. This will require separating the server initialization(invoking) and server definition logic into separate files.This way we can just import the httpServer and invoke it inside our tests and do other things like starting/stopping server instance within the tests/ randomizing the ports to avoid collision between tests etc.

@DMills27 @palisadoes @sumitra19jha @xoldyckk I think Instead of following only one approach for testing we should implement unit testing for functions that is independent of the database in integration testing we can mock data and test those functions.
in an integration test, we can use a mock database for testing resolvers mutation and query but integration testing will be slow as there is a database involved we should definitely do integration testing.
and for some cases, we should test resolvers thinking that it resolvers have been hited from the frontend for that we should considered an end to end testing considering using supertest or axios.

**Comment by palisadoes** [comment]
> > @ebrahim354
> > 
> > * What was the nature of the test you found with the resolver directly versus via the API?
> > * Could it have been discovered with more thorough testing via the API?
> > * If not, is it practical to use both approaches, or is it one or the other?
> > * The preferred approach is to have a standardized approach. Do you know of any way to make codecov work with APIs? Do you know of any other repos that have overcome this issue effectively
> > 
> > @xoldyckk It would be good to hear your thoughts on this.
> > This will impact the future development of the repo and we need to make the right decision now.
> 
> Since we are not mocking the database I think integration test corresponds the best with nature of resolver tests currently implemented. Though this is expected because the project is not following the best practices to easily implement unit tests. Business layer, service layer, database layer are all implented in the same scope. For example this is very easy to do in a Nest.js project because everything is injected into the child components through dependency injection and every individual component can easily be unit/integration tested at any level because those dependencies can be mocked or instantiated according to needs in testing environement. Since we're using apollo server, they've made it easy to do integration tests without invoking the whole http server through this approach :- [link](https://www.apollographql.com/docs/apollo-server/testing/testing/#testing-using-executeoperation). Though `i18n not initialized` error will kick in if we use this approach.
> 
> The nature of currently implemented API tests corresponds more to end to end testing.
> 
> It is very practical to use both kinds of approaches. For unit tests we use mock all internal/external services and test components in isolation. The integration test is done on a module(collection of files) which reflects a self-contained unit of our project. Something that doesn't depend on other modules and has every functionality it needs baked into it. For this we use mocks wherever possible and use real services if absolutely necessary. For end to end tests we simulate a real production-like environment inside our tests. Though it also depends on the size of our project, team, deadlines etc. It could be overkill to include every kind of test depending on those factors. I've been reading about testing in nodejs(server) ecosystem and I'm still confused what they mean by integration tests and e2e tests. Usually there's not a clear distinction. Some write tests making request to the running http server and call it integration test. Some call it e2e test. Honestly it makes more sense in the frontend world where there are specialized tools like `cypress` specifically for e2e tests.
> 
> The code coverage issue for API tests can be solved if the server we're testing against is invoked within the same process the test is running. This will require separating the server initialization(invoking) and server definition logic into separate files.This way we can just import the httpServer and invoke it inside our tests and do other things like starting/stopping server instance within the tests/ randomizing the ports to avoid collision between tests etc.

Based on your slack message in the `#talawa-api` channel, we should go with this approach. I confirmed this with @DMills27

**Comment by palisadoes** [comment]
To clarify, there is no intention to use a hosted version of Dgraph. The question was whether we could modify the API to have an easier way to do filtering so that we don't have so many resolvers. The Dgraph and other links below are used as an example of the syntax that could be considered.

- https://www.gatsbyjs.com/docs/query-filters/
- https://dgraph.io/docs/graphql/queries/and-or-not/#nesting

**Comment by ebrahim354** [comment]
the file is fully tested now so the issue should be closed, 
at least until there's a decision on what is the approach that will be taken in the future.
